### DISCUSSION TITLE
Cross Compilation Support

### DISCUSSION THREAD

**Discussion by pgm-n117** [NIKSS-vSwitch/nikss#111] [Q&A]
Hi devs,

I have been trying to use NIKSS on a Banana Pi R3, during my research on alternatives to use P4 on different hardware. I have managed to cross compile the library and binaries, to make it work on OpenWRT. I know this is not a specific question, but have you guys tried to execute this on ARM architechture based hardware? In my case, the cross compilation process for this device has gone well, but I'm encountering certain problems, maybe related to the OpenWRT kernel modules rather than the compilation process of NIKSS itself, but I wanted to share this experience, and know If you have any advice on this.

- In this case, when I try to load the simple-switch.p4 compiled program, I get the following warnings/errors:

``` 
root@OpenWrt:~# nikss-ctl pipeline load id 0 simple_switch.o
libbpf: elf: skipping unrecognized data section(10) .maps.clone_session_tbl_inner
libbpf: elf: skipping unrecognized data section(11) .maps.clone_session_tbl
libbpf: elf: skipping unrecognized data section(12) .maps.multicast_grp_tbl_inner
libbpf: elf: skipping unrecognized data section(13) .maps.multicast_grp_tbl
libbpf: elf: skipping unrecognized data section(14) .maps.ingress_tbl_fwd
libbpf: elf: skipping unrecognized data section(15) .maps.ingress_tbl_fwd_defaultAction
libbpf: elf: skipping unrecognized data section(16) .maps.hdr_md_cpumap
libbpf: elf: skipping unrecognized data section(17) .maps.crc_lookup_tbl
libbpf: elf: skipping unrecognized data section(34) .eh_frame
libbpf: elf: skipping relo section(35) .rel.eh_frame for section(34) .eh_frame
libbpf: Error in bpf_create_map_xattr(multicast_grp_tbl):No error information(-524). Retrying without BTF.
libbpf: Error in bpf_create_map_xattr(clone_session_tbl):No error information(-524). Retrying without BTF.
Pipeline id 0 successfully loaded!
```
- The result when adding a new port is the following:

``` 
root@OpenWrt:~# nikss-ctl add-port pipe 0 dev lan1
libbpf: Kernel error message: Underlying driver does not support XDP in native mode
XDP native mode not supported by driver, retrying with generic SKB mode
libbpf: Kernel error message: TC classifier not found
failed to attach bpf program to interface lan1: No such file or directory
failed to add port: No such file or directory
```
I don't understand why is reporting errors on TC, since I have been using TC configurations on other experiments.

- The report when showing the actual pipeline is the following (after adding a second port "lan2"):
``` 
root@OpenWrt:~# nikss-ctl pipeline show id 0
{
    "pipeline": {
        "id": 0,
        "load_time": "2024-06-06T11:47:10+0000",
        "bpf_hook": "TC",
        "has_egress_program": true,
        "ports": [
            {
                "name": "lan1",
                "port_id": 5
            },
            {
                "name": "lan2",
                "port_id": 6
            }
        ],
        "objects": [
            "ingress_tbl_fwd"
        ]
    }
```
- Finally, after installing the table rules to forward packets between two devices connected to "lan1" and "lan2", the table shows the following configuration, but ping between the devices is not working:
``` 
root@OpenWrt:~# nikss-ctl table get pipe 0 ingress_tbl_fwd
{
    "ingress_tbl_fwd": {
        "entries": [
            {
                "key": [
                    {
                        "type": "exact",
                        "value": "0xdca6325b73c2"
                    }
                ],
                "action": {
                    "id": 1,
                    "name": "ingress_do_forward",
                    "parameters": [
                        {
                            "name": "egress_port",
                            "value": "0x5"
                        }
                    ]
                },
                "DirectCounter": {},
                "DirectMeter": {}
            },
            {
                "key": [
                    {
                        "type": "exact",
                        "value": "0xdca632007933"
                    }
                ],
                "action": {
                    "id": 1,
                    "name": "ingress_do_forward",
                    "parameters": [
                        {
                            "name": "egress_port",
                            "value": "0x6"
                        }
                    ]
                },
                "DirectCounter": {},
                "DirectMeter": {}
            }
        ],
        "default_action": {
            "action": {
                "id": 0,
                "name": "_NoAction",
                "parameters": []
            },
            "DirectCounter": {},
            "DirectMeter": {}
        },
        "DirectCounter": {}
    }
```

**Comment by tatry** [comment]
We never tested other architectures than amd64 (x86_64), especially ARM.

> * In this case, when I try to load the simple-switch.p4 compiled program, I get the following warnings/errors:
>
>root@OpenWrt:~# nikss-ctl pipeline load id 0 simple_switch.o
libbpf: elf: skipping unrecognized data section(10) .maps.clone_session_tbl_inner
libbpf: elf: skipping unrecognized data section(11) .maps.clone_session_tbl
libbpf: elf: skipping unrecognized data section(12) .maps.multicast_grp_tbl_inner
libbpf: elf: skipping unrecognized data section(13) .maps.multicast_grp_tbl
libbpf: elf: skipping unrecognized data section(14) .maps.ingress_tbl_fwd
libbpf: elf: skipping unrecognized data section(15) .maps.ingress_tbl_fwd_defaultAction
libbpf: elf: skipping unrecognized data section(16) .maps.hdr_md_cpumap
libbpf: elf: skipping unrecognized data section(17) .maps.crc_lookup_tbl
libbpf: elf: skipping unrecognized data section(34) .eh_frame
libbpf: elf: skipping relo section(35) .rel.eh_frame for section(34) .eh_frame
libbpf: Error in bpf_create_map_xattr(multicast_grp_tbl):No error information(-524). Retrying without BTF.
libbpf: Error in bpf_create_map_xattr(clone_session_tbl):No error information(-524). Retrying without BTF.
Pipeline id 0 successfully loaded!

These "errors" are expected, so there is everything ok. This is just an inform that `libbpf` retries to load bpf map but with different configuration.

> * The result when adding a new port is the following:
>
>root@OpenWrt:~# nikss-ctl add-port pipe 0 dev lan1
libbpf: Kernel error message: Underlying driver does not support XDP in native mode
XDP native mode not supported by driver, retrying with generic SKB mode
libbpf: Kernel error message: TC classifier not found
failed to attach bpf program to interface lan1: No such file or directory
failed to add port: No such file or directory
>
>I don't understand why is reporting errors on TC, since I have been using TC configurations on other experiments.

NIKSS use, more or less, 3-4 bpf programs, each must be attached separately to an interface. This message might be related to some later bpf program, e.g. when attaching egress pipeline. Look [here for library implementation](https://github.com/NIKSS-vSwitch/nikss/blob/b8b1e829ce364d9caaf3ead9c37e376578edf287/lib/nikss_pipeline.c#L115-L133). Translating your output into more readable form would look like this:
```
Errors from attaching to ingress (basically, no errors):
libbpf: Kernel error message: Underlying driver does not support XDP in native mode
XDP native mode not supported by driver, retrying with generic SKB mode

Errors from attaching to egress (problem with TC?):
libbpf: Kernel error message: TC classifier not found
failed to attach bpf program to interface lan1: No such file or directory
failed to add port: No such file or directory
```
This problem is related to the next bullet point, listing ports attached to a pipeline. `nikss-ctl` looks only for ingress pipeline program (or XDP helper) and based on this lists interfaces. So, if egress pipeline failed to attach, interface still can be printed on the list. Look [here](https://github.com/NIKSS-vSwitch/nikss/blob/b8b1e829ce364d9caaf3ead9c37e376578edf287/lib/nikss_pipeline.c#L543-L563) and [here](https://github.com/NIKSS-vSwitch/nikss/blob/b8b1e829ce364d9caaf3ead9c37e376578edf287/lib/nikss_pipeline.c#L611-L620) for implementation.

Please note that on error(s) programs are not detached from interface when adding interface.

>Finally, after installing the table rules to forward packets between two devices connected to "lan1" and "lan2", the table shows the following configuration, but ping between the devices is not working:

Here I see two problems:
1. First already mentioned above. So please verify that ingress, egress and XDP programs are attached to interfaces.
2. Byte-order problem. Maybe somewhere there is missing or unexpected byte order conversion. So, you can try manually swap byte order for table entries. I don't know the width of your fields but there might be padding involved - you can try to add zeroes after or before value.

PS. I guess, as quick fix you can try to move whole program to ingress, leaving empty egress control block in your code. In this case egress program will be not generated by compiler and there would be no problem with attaching to egress...
